docs/adr/010_device_request_queue.md

```diff
@@ -0,0 +1,154 @@
+# Device Request Queue
+
+- **Feature**: [#1480]
+- **Status**: Proposed
+
+## Context
+
+LoRa devices, which can reach multiple LNSs, pose an interesting challenge for
+Edge Hub since Edge Hub wasn't designed for devices competing for gateways.
+When a device connection is made to IoT Hub, Edge Hub continues to maintain
+that connection in the open state (for network resiliency) until explicitly
+closed. Moreover, IoT Hub allows maintining only a single connection per
+device. When another is opened, the existing one is closed. If a device can
+reach multiple LNSs then Edge Hub and IoT Hub begin an aggressive ping-pong of
+connections being opened and closed. In other words, as each LNS or Edge Hub
+tries to open a connection for the same device, IoT Hub closes the connection
+for the other. But then the other tries to re-open the connection and the
+ping-pong continues. This problem needs addressing to avoid scalability
+limitations.
+
+The ping-pong can be pevented by electing a leader LNS for a device and
+announcing the decision to others using a C2D message. An LNS receiving such a
+message will then close that connection so that Edge Hub does't continue to
+maintain the connection in the open state through retires. The handling of the
+message needs to know:
+
+- It can close the connection right now, without interfering with a currently
+  running operation, such as a message delivery where the message could be
+  potentially dropped.
+
+- There are no operations starting around the same time that would cause it to
+  be opened again.
+
+The following cases can trigger operations from different threads:
+
+- An OTAA join message; `DeviceJoinLoader` eventually loads the twins and
+  stores them back up.
+
+- A data message; `DeviceLoaderSynchronizer` loads the twins for multiple
+  devices based on the `DevAddr` and adds them to the cache.
+
+- C2D message that needs to be delivered can trigger a load if the device is
```

Comment:
We are talking about Class C C2D messages here, correct? If yes, let's write it down please

Reply:
Clarified with a6e194f43784a80c4c8a96a6d242c75cd6e32a21.